### PR TITLE
ResolveImplicitTeam RPC

### PR DESCRIPTION
**Pull request by maxtaco**
- plumbed through with conflict info
- add a generic LRU that combined a memory LRU and also a DB-backed storage

### CODE REVIEW

File: go/libkb/assertion.go [hunk 1]
```diff
@@ -747,12 +747,12 @@ func ParseImplicitTeamDisplayNameSuffix(suffix string) (ret *keybase1.ImplicitTe
 	}
 
 	generation, err := strconv.Atoi(matches[2])
-	if err != nil || generation < 0 {
+	if err != nil || generation <= 0 {
```
Comment (maxtaco) [inline go/libkb/assertion.go]: setting a conflict suffix of `(conflict 2017-02-04 #0)` is never allowed.

**Comment by maxtaco** [comment]
ping @mlsteele !

File: go/systests/rpc_test.go [hunk 2]
```diff
@@ -516,6 +516,28 @@ func TestResolveIdentifyImplicitTeamWithDuplicates(t *testing.T) {
...
+		require.False(t, strings.Contains(res.Name, "conflicted"), "no conflicts")
+	} else {
+		require.True(t, strings.Contains(res.Name, "conflicted"), "found conflicted")
+		require.True(t, strings.Contains(res.Name, fmt.Sprintf("#%d", int(gen))), "found conflict gen #")
+	}
+}
+
+func doubleTestResolveImplicitTeam(t *testing.T, g *libkb.GlobalContext, id keybase1.TeamID, isPublic bool, gen keybase1.Seqno) {
+	testResolveImplicitTeam(t, g, id, isPublic, gen)
+	testResolveImplicitTeam(t, g, id, isPublic, gen)
```
Comment (mlsteele) [inline go/systests/rpc_test.go]: why 2 identical calls?

File: go/libkb/interfaces.go [hunk 3]
```diff
@@ -606,10 +606,31 @@ type TeamLoader interface {
...
 
+type LRUContext interface {
+	VLogContext
+	KVStoreContext
+	ClockContext
+}
+
+type LRUKeyer interface {
+	MemKey() string
+	DbKey() DbKey
```
Comment (mlsteele) [inline go/libkb/interfaces.go]: For a bit of convenience `MemKey` could be omitted from the interface and always use `DbKey().Key`.

File: go/teams/conflicts.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+		g.Log.CWarningf(ctx, "In fetching from cache: %s", err.Error())
+	}
+	if cv != nil {
+		if p, ok := cv.(*keybase1.ImplicitTeamConflictInfo); ok {
+			if p.IsConflict() {
+				ret.ConflictInfo = p
+			}
+			return ret, nil
+		}
+		g.Log.CWarningf(ctx, "Bad element of wrong type from cache: %T", cv)
```
Comment (mlsteele) [inline go/teams/conflicts.go]: That `cv` is an interface{} out here is unfortunate. Wrap LRUer in a type for conflict infos? Then `GetImplicitTeamConflictInfoCacher` would return that type

File: go/systests/rpc_test.go (hunk 2 above)
Comment (maxtaco) [reply go/systests/rpc_test.go]: the second one hits the cache

File: protocol/avdl/keybase1/teams.avdl [hunk 5]
```diff
@@ -643,7 +643,7 @@ protocol teams {
   }
 
   record ImplicitTeamConflictInfo {
-    int generation;
+    Seqno generation;
```
[the remark this answers is not in the document]
Comment (maxtaco) [reply protocol/avdl/keybase1/teams.avdl]: then we can make a new type

**Comment by maxtaco** [comment]
we really need to get this in, it's blocking KBFS work

File: go/teams/conflicts.go (hunk 4 above)
Comment (maxtaco) [reply go/teams/conflicts.go]: I don't think that's really necessary.  It's the same thing, we're going to need this cast somewhere.

**Pull request merged**